pkg/sql/show_create.go

```diff
@@ -263,11 +264,18 @@ func (p *planner) showCreateTable(
 
 // quoteNames quotes and adds commas between names.
 func quoteNames(names ...string) string {
-	nameList := make(tree.NameList, len(names))
-	for i, n := range names {
-		nameList[i] = tree.Name(n)
+	var f struct {
+		buf bytes.Buffer
+		ctx tree.FmtCtx
+	}
+	f.ctx = tree.MakeFmtCtx(&f.buf, tree.FmtSimple)
```

Comment:
Is this a pattern that everyone will need to start learning? I'm excited by this optimization but this seems like a pretty high mental cost. I don't really understand what this pattern is doing or why I should care - what is the point of `f`? we never use it directly, just its members.

Reply:
It forces the entire anonymous struct to escape to the heap in one allocation, instead of separate allocs for each field that will need to escape. I wouldn't say it's a pattern everyone needs to start learning, but the more people who are aware of it, the better. Still, it's obscure enough that it probably shouldn't be used outside of benchmark-directed optimization efforts like this one.

Reply:
I'm tempted to say that this kind of thing is obscure enough that we shouldn't bother doing it. If a new team member saw this (or me 6 months from now) they would probably assume this was a mistake or leftover from previous code structure and rightfully delete it.

Also, is this brittle with respect to changing versions of the Go language?

Unless this particular optimization has a major effect on GC behavior and performance, it seems questionable.

At the very least it *needs* a comment to tell people who come across it later and don't know about this optimization to leave this structure alone, as it has no semantic purpose.

Reply:
> Also, is this brittle with respect to changing versions of the Go language?

No, this behavior is pretty much required by the language spec. The anonymous struct is just like a named struct. If one of the members "escapes", the entire struct needs to be placed on the heap.

We could hide this optimization and not sprinkle it around the code:

```
func NewFmtCtxWithBuffer(flags FmtFlags) *FmtCtx {
  var f struct {
    buf bytes.Buffer
    ctx tree.FmtCtx
  }
  f.ctx = MakeFmtCtx(&f.buf, flags)
  return &f.ctx
}
```

Reply:
Great idea - @knz can you make this happen? And also add a comment above `f` so people don't delete it.

Reply:
👍 Centralizing this sounds like a good idea.